### PR TITLE
Story/vspc 116 2

### PR DESCRIPTION
**Pull request by prjadhav14**
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 
Adding provisions for staff user to add text blocks on space images and making those visible on staff site

...Put description here...
  
## Anything else the reviewer needs to know?

... describe here ...

### CODE REVIEW

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/SpaceTextBlockManager.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    @Autowired
+    private SpaceTextBlockDisplayRepository spaceTextBlockDisplayRepo;
+
+    @Override
+    public ISpaceTextBlockDisplay createTextBlock(String id, float positionX, float positionY, String text,
+            float height, float width){
+        ISpace source = spaceManager.getSpace(id);
+        ISpaceTextBlock textBlock=spaceTextBlockFactory.createSpaceTextBlock(text, source);
+        ISpaceTextBlockDisplay spaceTextBlockDisplay = spaceTextBlockDisplayFactory.createSpaceTextBlockDisplay(textBlock,positionX,positionY,height, width);
+        textBlock=spaceTextBlockRepo.save((SpaceTextBlock) textBlock);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/SpaceTextBlockManager.java:47]: `textBlock` isn't needed afterwards, is it? so no need to assign it here.

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/DeleteLinkController.java [hunk 2]
```diff
@@ -41,4 +45,10 @@
         externalLinkManager.deleteLink(linkId);
         return new ResponseEntity<String>(HttpStatus.OK);
     }
+
+    @RequestMapping(value = "/staff/space/{id}/textBlock/{blockId}", method = RequestMethod.DELETE)
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/DeleteLinkController.java:49]: deleting a text block shouldn't be in the DeleteLinkController

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
clickin on the save button doesn't do anything.

**Comment by jdamerow** [comment]
```
om.mysql.cj.jdbc.exceptions.MysqlDataTruncation: Data truncation: Data too long for column &#39;text&#39; at row 1
	com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:104)
	com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:955)
```

**Comment by prjadhav14** [comment]
> ```
> om.mysql.cj.jdbc.exceptions.MysqlDataTruncation: Data truncation: Data too long for column &#39;text&#39; at row 1
> 	com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:104)
> 	com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:955)
> ```

To resolve this we would need to run the below SQL scripts since String text is updated to @Lob which may not be taking effect since the table is already created in the environment.

drop table SpaceTextBlockDisplay;
drop table SpaceTextBlock;
commit;

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
- How can a text block be edited? (ah, that only seems to work after reloading the page)
- right now I can type into a text block after it has been created but that doesn't seem to do anything (it's not saved)
- on the staff side, the text block should show the actual text.
- the input field for the text needs to be bigger since it might hold quite a bit of text
- how can I delete a text block (same as above, only shows up after reloading the page, but the delete button only empties the form but doesn't seem to delete the text block)
- I can click on the image and the text block moves after I added it (but that doesn't change the actual text block)
- the text block should have a black or white lightly transparent background or something (talk to Adhvik how to make that nice looking)

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddSpaceTextBlockController.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+public class AddSpaceTextBlockController {
+
+    @Autowired
+    private ISpaceManager spaceManager;
+
+    @Autowired
+    private ISpaceTextBlockManager spaceTextBlockManager;
+
+
+    @RequestMapping(value = "/staff/space/{id}/textBlock", method = RequestMethod.POST)
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/AddSpaceTextBlockController.java:33]: let's lowercase `textblock`

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/DeleteSpaceTextBlockController.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import org.springframework.web.bind.annotation.RequestMethod;
+
+import edu.asu.diging.vspace.core.services.ISpaceTextBlockManager;
+
+@Controller
+public class DeleteSpaceTextBlockController {
+    @Autowired
+    private ISpaceTextBlockManager spaceTextBlockManager;
+
+    @RequestMapping(value = "/staff/space/{id}/textBlock/{blockId}", method = RequestMethod.DELETE)
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/DeleteSpaceTextBlockController.java:18]: same as above

File: vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditSpaceTextBlockController.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+@Controller
+public class EditSpaceTextBlockController{
+
+    @Autowired
+    private ISpaceManager spaceManager;
+
+    @Autowired
+    private ISpaceTextBlockManager spaceTextBlockManager;
+
+    @RequestMapping(value = "/staff/space/link/textBlock/{id}", method = RequestMethod.POST)
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/web/staff/EditSpaceTextBlockController.java:31]: same as above

File: vspace/src/main/webapp/WEB-INF/views/exhibition/space.html [hunk 6]
```diff
@@ -224,6 +224,25 @@
...
+    			var widthText = [[${block.width}]]*10;
+    			block = $('<span data-link-id="'+[[${block.id}]]+'" class="spaceLink-'+[[${block.id}]]+' Info_cz_Class"><textarea class="textBlock" style="height:'+heightText+'px; width:'+widthText+'px; resize:none;" readonly>'+[[${block.spaceTextBlock.text}]]+'</textarea></span>');
+    			block.css('position', 'absolute');
+    			var height = parseInt($("#bgImage").css("height"));
+            	var width = parseInt($("#bgImage").css("width"));
+            	var	linkPosX=(width*[[${block.positionX}]])/[[${display.width}]];
+            	var	linkPosY=(height*[[${block.positionY}]])/[[${display.height}]];
+    			block.css('left', linkPosX + posX);
+    			block.css('top', linkPosY + posY);
+    			$("#space").append(block);
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/exhibition/space.html:243]: indentations

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
- textbox can't be changed after creation when using the corner (not the input fields)
- when using the input fields, there is something off. When I change the width and then the hieght, the width gets reset or something (and the other way around)
- I can change the size of the textbox by dragging the corner but that actually changes the textbox. I shouldn't be able to change the textbox if it doesn't actually change the size (hope that makes sense).

**Comment by jdamerow** [comment]
Resolve conflicts please

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
- the "edit text block" should be "Save" and the buttons need to have a little bit of padding at the bottom:
image
- the sizing is different on staff and public page:
staff:
image
public:
image
- resizing the public page window does this:
image

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
staff side:
- when editing a text block: I click on the text block, it opens the edit modal. when I click somewhere else to move the text block, it duplicates it.
- when creating a text block, the spinning icon is not shown. It just takes a long time for the modal to close, which is confusing since the user doesn't know if it didn't work or if it is just slow.
- on the main image (not the modal), the right bottom corner shouldn't let the user change the size of the box since that doesn't actually change the size of the box.

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceTextBlockManagerTest.java [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        
+        Mockito.when(spaceTextBlockRepo.findById("SPB001")).thenReturn(mockSpaceTextBlock);
+        
+        managerToTest.deleteTextBlock("SPB001");
+        
+        Mockito.verify(spaceTextBlockDisplayRepo).deleteBySpaceTextBlock(mockSpaceTextBlock.get());
+        Mockito.verify(spaceTextBlockRepo).delete(mockSpaceTextBlock.get());
+    }
+    
+
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceTextBlockManagerTest.java:175]: test cases for not succeeding are missing (e.g. if text block to be deleted doesn't exist).

**Review by jdamerow — CHANGES_REQUESTED** [review]
CodeFactor also found an issue.

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html [hunk 8]
```diff
@@ -686,7 +816,43 @@
...
+function getTextArea(x,y) {
+    var posX = $("#bgImage").position().left;
+    var posY = $("#bgImage").position().top;
+    var text = $('<span id="text"><textarea id="textB" name="textBlock" type="textarea" class="textBlock" style="resize:both;"></textarea></span>');
+    text.css('position', 'absolute');
+    text.css('left', x + posX);
+    text.css('top', y);
+    $("#space").append(text);
+    
+    var textB = document.getElementById('textB'),
```
[the remark this answers is not in the document]
Comment (jdamerow) [reply vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html:854]: ?

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html [hunk 9]
```diff
@@ -686,7 +841,43 @@
...
+    tHeight = textB.clientHeight,
+    tWidth = textB.clientWidth;
+    
+    var height = parseInt($("#bgImage").css("height"));
+    var width = parseInt($("#bgImage").css("width"));
```
[the remark this answers is not in the document]
Comment (jdamerow) [reply vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html:859]: ?

**Review by jdamerow — CHANGES_REQUESTED** [review]
there are unaddressed comments

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceTextBlockManagerTest.java [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+        Assert.assertEquals(spaceTextBlockDisplay.getName(), updatedSpaceTextBlockDisplay.getName());
+        Assert.assertEquals(new Double(spaceTextBlockDisplay.getPositionX()), new Double(updatedSpaceTextBlockDisplay.getPositionX()));
+        Assert.assertEquals(new Double(spaceTextBlockDisplay.getPositionY()), new Double(updatedSpaceTextBlockDisplay.getPositionY()));
+        Assert.assertEquals(new Float(spaceTextBlockDisplay.getWidth()), new Float(updatedSpaceTextBlockDisplay.getWidth()));
+        Assert.assertEquals(new Float(spaceTextBlockDisplay.getHeight()), new Float(updatedSpaceTextBlockDisplay.getHeight()));
+        Assert.assertEquals(spaceTextBlockDisplay.getSpaceTextBlock(), updatedSpaceTextBlockDisplay.getSpaceTextBlock());
+    }
+    
+    @Test
+    public void test_updateTextBlock_textblockIdempty_failed(){
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceTextBlockManagerTest.java:125]: fix camel case

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceTextBlockManagerTest.java [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+        Optional<SpaceTextBlockDisplay> mockSpaceTextBlockDisplay = Optional.of(spaceTextBlockDisplay);
+        
+        Mockito.when(spaceTextBlockRepo.findById(spaceTextBlock.getId())).thenReturn(Optional.empty());
+        Mockito.when(spaceTextBlockDisplayRepo.findById(spaceTextBlockDisplay.getId())).thenReturn(mockSpaceTextBlockDisplay);
+        
+        Mockito.verify(spaceTextBlockRepo, Mockito.never()).save(spaceTextBlock);
+    }
+    
+    @Test
+    public void test_updateTextBlock_textblocDisplayIdempty_failed(){
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceTextBlockManagerTest.java:148]: fix camel case and spelling  mistake

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceTextBlockManagerTest.java [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+        spaceTextBlockDisplay.setSpaceTextBlock(spaceTextBlock);
+        
+        Optional<SpaceTextBlock> mockSpaceTextBlock = Optional.of(spaceTextBlock);
+        Optional<SpaceTextBlockDisplay> mockSpaceTextBlockDisplay = Optional.of(spaceTextBlockDisplay);
+        
+        Mockito.when(spaceTextBlockRepo.findById(spaceTextBlock.getId())).thenReturn(Optional.empty());
+        Mockito.when(spaceTextBlockDisplayRepo.findById(spaceTextBlockDisplay.getId())).thenReturn(mockSpaceTextBlockDisplay);
+        
+        Mockito.verify(spaceTextBlockRepo, Mockito.never()).save(spaceTextBlock);
+    }
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceTextBlockManagerTest.java:145]: this method never calls `managerToTest`, does it?

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceTextBlockManagerTest.java [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+        spaceTextBlockDisplay.setWidth(40);
+        spaceTextBlockDisplay.setHeight(50);
+        spaceTextBlockDisplay.setSpaceTextBlock(spaceTextBlock);
+        
+        Optional<SpaceTextBlock> mockSpaceTextBlock = Optional.of(spaceTextBlock);
+        
+        Mockito.when(spaceTextBlockRepo.findById(spaceTextBlock.getId())).thenReturn(mockSpaceTextBlock);
+        Mockito.when(spaceTextBlockDisplayRepo.findById(spaceTextBlockDisplay.getId())).thenReturn(Optional.empty());
+        
+        Mockito.verify(spaceTextBlockDisplayRepo, Mockito.never()).save(spaceTextBlockDisplay);
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SpaceTextBlockManagerTest.java:166]: same here, never calls `managerToTest`

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html [hunk 14]
```diff
@@ -191,13 +196,53 @@
...
+                var blockData = $('#'+spanId+block.spaceTextBlock.id).data();
+                makeTextBlockEditable(blockData["blockid"],blockData["linkId"],blockData["text"],block.positionX,block.positionY,blockData["width"],blockData["height"]);
+            });
+        }
+        
+        if(block.spaceTextBlock.id === selectedTextBlockId) {
+        	var textModalBlock = $("#textModalBlockId-"+selectedTextBlockId).find("textarea")[0];
+        	
+        	tHeight = textModalBlock.clientHeight,
+            tWidth = textModalBlock.clientWidth;
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html:218]: indent

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html [hunk 15]
```diff
@@ -191,13 +196,53 @@
...
+
 function drawLinks(){
     $(".Info_cz_Class").remove();
     $(".imageLink").remove();
     $(".alertLink").remove();
     $(".ellipse_class").remove();
     $(".unpublishedSpaceClass").remove();
-    $(".deletedSpaceClass").remove();
+    $(".textBlockClass").remove();
+	$(".deletedSpaceClass").remove();
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html:245]: indent

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html [hunk 16]
```diff
@@ -610,6 +741,8 @@
     drawLinks();
     $("#space").css("display", "-webkit-box");
     $(formToShow).css("display", "flex");
+    modalBgImageHeight = parseInt($("#bgImage").css("height"));
+	modalBgImageWidth = parseInt($("#bgImage").css("width"));
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.html:745]: indent

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Comment by jdamerow** [comment]
There is a weird bug on the public site. I think the text in the textbox is also printed behind the box and doesn't scroll there.
image

**Comment by tusanand** [comment]
Fixed the bug where the text blocks were loading twice on exhinition page.
Fixed all indentation and Unit test case issues

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Pull request merged**